### PR TITLE
IGNITE-22771 Create new System configuration root

### PR DESCRIPTION
**Pull request by Cyrill**
https://issues.apache.org/jira/browse/IGNITE-22771

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            return ensureCreated ? ensureCreated(resolved) : resolved;
+        });
+    }
+
+    @Override
+    public Path get() {
+        return super.get();
+    }
+
+    private static Path ensureCreated(Path dir) {
```
Comment (alievmirza) [inline modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java:123]: Why do we have two the same methods here and in `org.apache.ignite.internal.configuration.IgnitePaths#ensureCreated`

File: modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+     * Resolve the other path against this one.
+     */
+    public LazyPath resolveLazy(Path other) {
+        return resolveLazy(other, false);
+    }
+
+    /**
+     * Resolve the other path against this one.
+     */
+    public LazyPath resolveLazy(String other) {
```
Comment (alievmirza) [inline modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java:92]: these two methods are just the same

File: modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        return new LazyPath(() -> pathOrDefault(pathSupplier.get(), defaultPath));
+    }
+
+    /**
+     * Create a new instance from {@code pathSupplier} if it returns a nonempty value, otherwise use {@code defaultPathSupplier}.
+     *
+     * @param pathSupplier Path supplier.
+     * @param defaultPathSupplier Default path supplier.
+     */
+    public static LazyPath create(Supplier<String> pathSupplier, Supplier<Path> defaultPathSupplier) {
```
Comment (alievmirza) [inline modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java:62]: this method is not used

File: modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        return new LazyPath(() -> defaultPath);
+    }
+
+    /**
+     * Create a new instance from {@code pathSupplier} if it returns a nonempty value, otherwise use {@code defaultPath}.
+     *
+     * @param pathSupplier Path supplier.
+     * @param defaultPath Default path.
+     */
+    public static LazyPath create(Supplier<String> pathSupplier, Path defaultPath) {
```
Comment (alievmirza) [inline modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java:52]: this method is not used

File: modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java (hunk 1 above)
Comment (alievmirza) [inline modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java:123]: I don't understand scenarios when we nedd ensured or not ensured paths

File: modules/system-configuration/src/main/java/org/apache/ignite/internal/configuration/SystemConfigurationSchema.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+import org.apache.ignite.configuration.annotation.ConfigurationRoot;
+import org.apache.ignite.configuration.annotation.ConfigurationType;
+import org.apache.ignite.configuration.annotation.Value;
+
+/**
+ * System Configuration schema.
+ */
+@ConfigurationRoot(rootName = "system", type = ConfigurationType.LOCAL)
+public class SystemConfigurationSchema {
```
Comment (alievmirza) [inline modules/system-configuration/src/main/java/org/apache/ignite/internal/configuration/SystemConfigurationSchema.java:28]: I'm not sure that this is a good naming. I would name it like 'PathsConfigurationSchema`. As I understood, this class will have other paths, not only `partitionsBasePath`. The same for root, I would name it `paths`

File: modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    @Override
+    public Path get() {
+        return super.get();
+    }
+
+    private static Path ensureCreated(Path dir) {
+        try {
+            Files.createDirectories(dir);
+        } catch (IOException e) {
+            throw new IgniteInternalException("Failed to create directory: " + dir + ": " + e.getMessage(), e);
```
Comment (alievmirza) [inline modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java:127]: deprecated constructor usage

**Review by alievmirza — CHANGES_REQUESTED** [review]

File: modules/system-configuration/src/main/java/org/apache/ignite/internal/configuration/SystemConfigurationSchema.java (hunk 5 above)
Comment (sashapolo) [reply modules/system-configuration/src/main/java/org/apache/ignite/internal/configuration/SystemConfigurationSchema.java:28]: I agree with Mirza on this, "system" sounds too vague in this context

File: modules/system-configuration/src/main/java/org/apache/ignite/internal/configuration/SystemConfigurationSchema.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Directory where partition data is stored. By default "partitions" subfolder of data storage path is used.
+     *
+     * <pre>
+     * The internal directory structure contains:
```
Comment (sashapolo) [inline modules/system-configuration/src/main/java/org/apache/ignite/internal/configuration/SystemConfigurationSchema.java:34]: I've heard that we wanted to be able to have raft log and partition storages on separate devices, for example. How will this configuration allow that?

Comment (sashapolo) [reply modules/system-configuration/src/main/java/org/apache/ignite/internal/configuration/SystemConfigurationSchema.java:34]: This ticket is about decoupling MS and CMG Raft logs from partitions. I'm talking about separating partitions raft logs and partitions data

Comment (sashapolo) [reply modules/system-configuration/src/main/java/org/apache/ignite/internal/configuration/SystemConfigurationSchema.java:34]: The current javadoc is misleading now. We have `partitionsBasePath` that is responsible for the log, meta and db files, while we also have `partitionsLogPath` which is also responsible for the log files? Why do we have this overlapping? Can we have three properties each responsible for its own folder?

File: modules/system-configuration/build.gradle [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    annotationProcessor project(':ignite-configuration-annotation-processor')
+    annotationProcessor libs.auto.service
+
+    implementation project(':ignite-core')
+    implementation project(':ignite-configuration-api')
+    implementation libs.jetbrains.annotations
+    implementation libs.auto.service.annotations
+}
+
+description = 'ignite-system-configuration'
```
Comment (ptupitsyn) [inline modules/system-configuration/build.gradle:33]: Do we really need a separate module just for configuration?

Comment (ptupitsyn) [reply modules/system-configuration/build.gradle:33]: Seems weird to have a module just for that, can we put it into an existing module maybe?

Comment (ptupitsyn) [reply modules/system-configuration/build.gradle:33]: Can it be `ignite-runner` or `ignite-configuration`?

File: modules/partition-replicator/build.gradle [hunk 9]
```diff
@@ -74,6 +74,7 @@ dependencies {
     integrationTestImplementation project(':ignite-page-memory')
     integrationTestImplementation project(':ignite-placement-driver')
     integrationTestImplementation project(':ignite-affinity')
+    integrationTestImplementation project(':ignite-runner')
```
Comment (ascherbakoff) [inline modules/partition-replicator/build.gradle:77]: What is the purpose of this dependency ?

**Review by sashapolo — APPROVED** [review]

**Review by ascherbakoff — APPROVED** [review]

**Review by ptupitsyn — APPROVED** [review]

**Pull request merged**